### PR TITLE
Throw MisconfigurationError on unknown mode

### PR DESCRIPTION
**Pull request by alanhdu**
Potential closes  https://github.com/PyTorchLightning/pytorch-lightning/issues/5252

When an unknown `mode` is passed to `EarlyStopping` or `ModelCheckpoitn`, raise a misconfiguration error instead of just a runtime warning.

### CODE REVIEW

**Review by ydcjeff — COMMENTED** [review]
Would you mind add a test for that?

**Comment by alanhdu** [comment]
@ydcjeff: Done (sorry for the delay -- I was out during the holidays)!

Rebased onto the latest master as well.

File: tests/callbacks/test_early_stopping.py [hunk 1]
```diff
@@ -294,3 +294,27 @@ def validation_epoch_end(self, outputs):
...
+def test_model_checkpoint_mode_options():
+    with pytest.raises(MisconfigurationException) as exc_info:
+        ModelCheckpoint(mode="unknown_option")
+    assert "unknown_option" in str(exc_info.value)
+
+    with pytest.warns(DeprecationWarning) as exc_info:
+        ModelCheckpoint()
+
+    with pytest.warns(DeprecationWarning):
+        ModelCheckpoint(mode="auto")
```
Comment (Borda) [inline tests/callbacks/test_early_stopping.py:320]: pls add also a matching pattern, so we are clear what error it is...

Comment (alanhdu) [reply tests/callbacks/test_early_stopping.py]: Done!

**Review by awaelchli — APPROVED** [review]
nice.
a minor improvement would be to print available modes in the misconfiguration exception, i.e., 
mode x is unknown. Available modes are min, max.

**Review by ydcjeff — COMMENTED** [review]
these tests can be removed?

File: tests/callbacks/test_early_stopping.py [hunk 2]
```diff
@@ -294,3 +294,25 @@ def validation_epoch_end(self, outputs):
...
+
+def test_early_stopping_mode_options():
+    with pytest.raises(MisconfigurationException, match="unknown_option"):
+        EarlyStopping(mode="unknown_option")
+
+    with pytest.warns(DeprecationWarning, match="'auto' is deprecated"):
+        EarlyStopping()
+
+    with pytest.warns(DeprecationWarning, match="'auto' is deprecated"):
+        EarlyStopping(mode="auto")
```
Comment (ydcjeff) [inline tests/callbacks/test_early_stopping.py:307]: added in
https://github.com/PyTorchLightning/pytorch-lightning/blob/410d67fbe866ee20069e88db1be729d27ae0af48/tests/deprecated_api/test_remove_1-3.py#L40-L41

File: tests/callbacks/test_early_stopping.py [hunk 3]
```diff
@@ -294,3 +294,25 @@ def validation_epoch_end(self, outputs):
...
+
+def test_model_checkpoint_mode_options():
+    with pytest.raises(MisconfigurationException, match="unknown_option"):
+        ModelCheckpoint(mode="unknown_option")
+
+    with pytest.warns(DeprecationWarning, match="'auto' is deprecated"):
+        ModelCheckpoint()
+
+    with pytest.warns(DeprecationWarning, match="'auto' is deprecated"):
+        ModelCheckpoint(mode="auto")
```
Comment (ydcjeff) [inline tests/callbacks/test_early_stopping.py:318]: added in
https://github.com/PyTorchLightning/pytorch-lightning/blob/410d67fbe866ee20069e88db1be729d27ae0af48/tests/deprecated_api/test_remove_1-3.py#L37-L38

**Review by ydcjeff — COMMENTED** [review]
These tests can be removed not to collide with deprecated ones?

**Comment by Borda** [comment]
@alanhdu mind rebase on `PyTorchLightning:release/1.2-dev` 🐰

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/callbacks/early_stopping.py [hunk 4]
```diff
@@ -96,15 +97,10 @@ def __init__(
...
     def __init_monitor_mode(self):
-        # TODO: Update with MisconfigurationException when auto mode is removed in v1.3
         if self.mode not in self.mode_dict and self.mode != 'auto':
-            if self.verbose > 0:
-                rank_zero_warn(
-                    f'EarlyStopping mode={self.mode} is unknown, fallback to auto mode.',
-                    RuntimeWarning,
-                )
-            self.mode = 'auto'
+            raise MisconfigurationException(f'EarlyStopping mode={self.mode} is unknown.')
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/early_stopping.py:101]:
```suggestion
            raise MisconfigurationException(
                f"`mode` can be auto, {', '.join(self.mode_dict.keys())}, got {mode}"
            )
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 5]
```diff
@@ -289,14 +289,10 @@ def __init_monitor_mode(self, monitor, mode):
...
         }
 
-        # TODO: Update with MisconfigurationException when auto mode is removed in v1.3
         if mode not in mode_dict and mode != 'auto':
-            rank_zero_warn(
-                f"ModelCheckpoint mode {mode} is unknown, fallback to auto mode",
-                RuntimeWarning,
-            )
-            mode = "auto"
+            raise MisconfigurationException(f'ModelCheckpoint mode={mode} is unknown.')
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:293]:
```suggestion
            raise MisconfigurationException(
                f"`mode` can be auto, {', '.join(mode_dict.keys())}, got {mode}"
            )
```

File: tests/callbacks/test_early_stopping.py [hunk 6]
```diff
@@ -297,3 +297,13 @@ def validation_epoch_end(self, outputs):
...
     _logger.disabled = False
+
+
+def test_early_stopping_mode_options():
+    with pytest.raises(MisconfigurationException, match="unknown_option"):
+        EarlyStopping(mode="unknown_option")
+
+
+def test_model_checkpoint_mode_options():
+    with pytest.raises(MisconfigurationException, match="unknown_option"):
```
Comment (rohitgr7) [inline tests/callbacks/test_early_stopping.py:308]:
```suggestion
    with pytest.raises(MisconfigurationException, match="`mode` can be auto, .* got unknown_option"):
```

File: tests/callbacks/test_early_stopping.py [hunk 7]
```diff
@@ -297,3 +297,13 @@ def validation_epoch_end(self, outputs):
         (trainer.global_step, max(min_steps, by_early_stopping, by_min_epochs), step_freeze, min_steps, min_epochs)
 
     _logger.disabled = False
+
+
+def test_early_stopping_mode_options():
+    with pytest.raises(MisconfigurationException, match="unknown_option"):
```
Comment (rohitgr7) [inline tests/callbacks/test_early_stopping.py:303]:
```suggestion
    with pytest.raises(MisconfigurationException, match="`mode` can be auto, .* got unknown_option"):
```

**Review by rohitgr7 — APPROVED** [review]
nice!

**Comment by alanhdu** [comment]
Rebased to fix the merge conflict.

**Review by justusschock — APPROVED** [review]

**Pull request merged**